### PR TITLE
Update filesystem datastore to async/await

### PR DESCRIPTION
**Pull request by ntninja**
* Update filesystem datastore to async/await
 * Slightly restructure the `datastore` API again, everything important should now be exposed and `datastore.core.*` only needed by internal stuff

### CODE REVIEW

File: datastore/core/test/conftest.py
Comment (AliabbasMerchant) [inline datastore/core/test/conftest.py:38]: Nice!

File: datastore/__init__.py
Comment (AliabbasMerchant) [inline datastore/__init__.py:13]: Nice work exposing those Classes directly.

File: datastore/__init__.py
Comment (AliabbasMerchant) [inline datastore/__init__.py:34]: We will also have to expose FileSystemDatastore and FileReader

File: datastore/filesystem/filesystem.py
Comment (AliabbasMerchant) [inline datastore/filesystem/filesystem.py:61]: Is it necessary to call .aclose() here? Won't the user explicitly call it?
And if we are already calling .aclose(), when the user explicitly calls .aclose(), wont there be some kind of error, saying that it is already closed?

File: datastore/filesystem/filesystem.py
Comment (AliabbasMerchant) [inline datastore/filesystem/filesystem.py]: I didn't understand the logic here, so no comments

**Comment by AliabbasMerchant** [comment]
@alexander255 Sorry for the delay in reviewing.

I went through all the files. They look good to me.  
(3 out of 56 code tests are failing. [TestQuery.test_cursor, TestOrder.test_basic, TestFilter.test_basic]  And these are from broken modules, right?  
Also, can you please try fixing Travis?)  

Some of the code snippets in the doc comments have incorrect imports. But I think we can fix them later on.

Can we, in a future commit, shift all the test files to a separate folder outside?

Please correct me if I am wrong:  
1. You are exposing Abstract Classes via datastore.abc
2. And exposing the utils via datastore.utils
3. mypy is a dependency because of the typing and all.

**Review by AliabbasMerchant — APPROVED** [review]

File: datastore/__init__.py
Comment (ntninja) [reply datastore/__init__.py:34]: My thought was that the filesystem datastore has to be explicitly imported using `datastore.filesystem` as it is not needed if another datastore (like leveldb or the in-memory dict one) is used.

`FileReader` is just an implementation detail of how the datastore manages to return its `datastore.abc.ReceiveStream`. Its not really meant to be used directly.

File: datastore/filesystem/filesystem.py
Comment (ntninja) [reply datastore/filesystem/filesystem.py:61]: Calling `.aclose()` here ensures that the file automatically be closed in an

```py
async for chunk in await datastore.get("whatever"):
    …
```

kind of scenario. Otherwise the calling code would have to look like this:

```py
async with await datastore.get("whatever") as file:
    async for chunk in file:
        …
```

… or the file descriptor would leak.

Calling `aclose` several times [is not an error](https://trio.readthedocs.io/en/stable/reference-io.html#trio.abc.AsyncResource.aclose):

> If the resource is already closed, then this method should silently succeed.

File: datastore/filesystem/filesystem.py
Comment (ntninja) [reply datastore/filesystem/filesystem.py]: It tries to `rmdir` parent directories as long as

  1. that directory is a subdirectory of the `root_path`
     * that is: the directory path starts with the string `<…/some/root/path>/`
  2. and the directory's parent is not the directory itself again
     * this guards against the case of the root directory being the current directory (`.`) whose `.parent` property is also `.` (I'm am not sure anymore if this isn't subsumed the first rule however as `".".startswith("./")` is `False` either way…?

I probably should a comment to the code about this, shouldn't I?

**Comment by adamjsawicki** [comment]
@AliabbasMerchant I fixed travis problems in alexander255/py-datastore#1. @alexander255 made changes you'd suggested.

**Comment by ntninja** [comment]
@AliabbasMerchant: Thank you for your review, better late than never I guess! :slightly_smiling_face: 

> I went through all the files. They look good to me.
> (3 out of 56 code tests are failing. [TestQuery.test_cursor, TestOrder.test_basic, TestFilter.test_basic] And these are from broken modules, right?

I think so, but they should be skipped in the test suite. I created #14 to document what I think `query` will be in the future. After doing that there should be no more failing tests. Note that that work is **not** py-ipfs related, it's just something to make this library complete again.

> Also, can you please try fixing Travis?)

@Swixx did that but it's waiting for this to be merged.

> Some of the code snippets in the doc comments have incorrect imports. But I think we can fix them later on.

Created #16 about this.

> Can we, in a future commit, shift all the test files to a separate folder outside?

Sure, now that the tests (almost) only use the public API, I seeno reason to keep them in those folders.

> Please correct me if I am wrong:
>
> 1. You are exposing Abstract Classes via datastore.abc
> 2.  And exposing the utils via datastore.utils

Yes to both.

> 3.  mypy is a dependency because of the typing and all.

No, mypy is only a dependency for type checking. It is not used during testing or in the final package otherwise. (However the stdlib `typing` module is used a lot for proper typing declarations in all sorts of places, making it a hard dependency.)

If you are referring to [this change](https://github.com/ipfs/py-datastore/pull/13/commits/e80a1d731eff2737695161538f0e663036e88545#diff-b91f3d5bd63fcd17221b267e851608e8R39). This is about forcing mypy to be downloaded in source form, so that [mypy_run.py](https://github.com/ipfs/py-datastore/blob/master/mypy_run.py) can monkey patch the use of some extra typing definitions into the mypy validator. (The previous form was incorrect and caused issues.)

**Comment by AliabbasMerchant** [comment]
Great!  

> If you are referring to this change. This is about forcing mypy to be downloaded in source form, so that mypy_run.py can monkey patch the use of some extra typing definitions into the mypy validator. (The previous form was incorrect and caused issues.)

Yes, I was referring to that change. Got it 👍 

Lets merge this!

Then we can merge the Travis part that @Swixx did, and #16  
And then `query` and shifting of tests to a separate folder.

**Comment by ntninja** [comment]
@AliabbasMerchant: Ok, I'll merge this once I added that comment to the filesystem.py source file. (Also I realized that the `rmdir` should be optional entirely.)

Probably later this evening or tomorrow.

**Comment by AliabbasMerchant** [comment]
Okay.  
Also, @alexander255, can you please tell the ipfs admins, to include this repo under the python team? (I am not aware of the procedure to be followed to request this.)

File: datastore/filesystem/filesystem.py
Comment (AliabbasMerchant) [inline datastore/filesystem/filesystem.py]: Got it! Thanks!  
It was a wise decision to use pathlib 👍

File: datastore/filesystem/filesystem.py
Comment (AliabbasMerchant) [inline datastore/filesystem/filesystem.py:325]: Got it! Thanks!
It was a wise decision to use pathlib 👍

File: datastore/__init__.py
Comment (AliabbasMerchant) [reply datastore/__init__.py:34]: Okay 👍

**Pull request merged**
